packages/alfa-aria/src/node.ts

```diff
@@ -222,7 +223,65 @@ export namespace Node {
 
     const root = node.root({ flattened: true });
 
-    build(root, device);
+    // Before we start constructing the accessibility tree, we need to resolve
+    // explicit ownership of elements as specified by the `aria-owns` attribute.
+    // https://w3c.github.io/aria/#aria-owns
+
+    // Find all elements in the tree. As explicit ownership is specified via ID
+    // references, it cannot cross shadow or document boundaries.
+    const elements = root.inclusiveDescendants().filter(dom.Element.isElement);
+
+    // Build a map from ID -> element to allow fast resolution of ID references.
+    // The collected references are added to the map in reverse order to ensure
+    // that the first occurrence of a given ID is what ends up in the map in
+    // event of duplicates.
+    const ids = Map.from(
+      elements
+        .collect((element) => element.id.map((id) => [id, element] as const))
+        .reverse()
+    );
+
+    // Do a first pass over `aria-owns` attributes and collect the referenced
+    // elements.
+    const references = elements.collect((element) =>
+      element.attribute("aria-owns").map(
+        (attribute) =>
+          [
+            element,
+            attribute
+              .tokens()
+              .collect((id) => ids.get(id))
+              // Reject select-references.
+              .reject((reference) => reference === element),
+          ] as const
+      )
+    );
+
+    // Refine the collected `aria-owns` references, constructing a set of
+    // claimed elements and resolving conflicting claims as needed.
+    const [claimed, owned] = references.reduce(
+      ([claimed, owned], [element, references]) => {
+        // Reject all element references that have already been claimed.
+        references = references.reject((element) => claimed.has(element));
+
+        // If there are no references left, this element has no explicit
+        // ownership.
+        if (references.isEmpty()) {
+          return [claimed, owned];
+        }
+
+        // Claim the remaining references.
+        claimed = references.reduce(
+          (claimed, element) => claimed.add(element),
+          claimed
+        );
+
+        return [claimed, owned.set(element, references)];
+      },
+      [Set.empty<dom.Node>(), Map.empty<dom.Element, Sequence<dom.Node>>()]
+    );
+
+    build(root, device, claimed, owned);
 
     if (_cache.has(node)) {
       return _cache.get(node).get();
```

Comment:
Shouldn't we look in the cache first? If the node is already there, no need to process `aria-owns` again…

Reply:
D'oh 🤦

Reply:
Oh wait, I forgot that it already does 🤦 https://github.com/Siteimprove/alfa/blob/7d16fa50924bdf67db8ae53c878a27ad99e6223b/packages/alfa-aria/src/node.ts#L218-L222

Reply:
We should however definitely cache the `Inert` node returned if the node isn't even part of the tree.